packages/grid/x-data-grid/src/hooks/features/rows/useGridVisibleRowsState.ts

```diff
@@ -0,0 +1,45 @@
+import * as React from 'react';
+import { GridPrivateApiCommunity } from '../../../models/api/gridApiCommunity';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import {
+  GRID_DEFAULT_STRATEGY,
+  GridStrategyProcessor,
+  useGridRegisterStrategyProcessor,
+} from '../../core/strategyProcessing';
+import type { GridStateInitializer } from '../../utils/useGridInitializeState';
+
+export const visibleRowsStateInitializer: GridStateInitializer = (state) => {
+  return {
+    ...state,
+    visibleRows: {
+      lookup: {},
+    },
+  };
+};
+
+export function useGridVisibleRowsState(apiRef: React.MutableRefObject<GridPrivateApiCommunity>) {
+  const updateVisibleRowsState = React.useCallback(() => {
+    apiRef.current.setState((state) => {
+      const visibleRowsState = apiRef.current.applyStrategyProcessor('visibleRows', {
+        tree: state.rows.tree,
+        filteredRowsLookup: state.filter.filteredRowsLookup,
+      });
+      return {
+        ...state,
+        visibleRows: visibleRowsState,
+      };
+    });
+    apiRef.current.forceUpdate();
+  }, [apiRef]);
+
+  const getVisibleRows = React.useCallback<GridStrategyProcessor<'visibleRows'>>((params) => {
+    // For flat tree, the `visibleRowsLookup` and the `filteredRowsLookup` are equals since no row is collapsed.
+    return {
+      lookup: params.filteredRowsLookup,
+    };
+  }, []);
```

Comment:
I know the code style is to inline these everywhere, but this is a pure function that doesn't need to be a closure. It could be extracted outside this hook, which would avoid the repeated allocations & `useCallback` costs.

Sidenote, but I feel like the current code/style conventions aren't optimal for performance, we allocate a lot of stuff.

Reply:
Good point, done :+1: